googletest/feature_support_test.cpp

```diff
@@ -0,0 +1,1313 @@
+{
+    device->SetNodeCount(3);
+    device->m_ProtectedResourceSessionTypeCount = {3, 14, 21};
+    device->m_ProtectedResourceSessionTypes[0].resize(3);
+    device->m_ProtectedResourceSessionTypes[1].resize(14);
+    device->m_ProtectedResourceSessionTypes[2].resize(21);
+
+    INIT_FEATURES();
+
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(0), 3);
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(1), 14);
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(2), 21);
+}
+
+// Unavailable Test
+TEST_F(FeatureSupportTest, ProtectedResourceSessionTypeCountUnavailable)
+{
+    device->m_ProtectedResourceSessionTypeCountAvailable = false;
+    device->SetNodeCount(3);
+    device->m_ProtectedResourceSessionTypeCount = {3, 14, 21};
+    device->m_ProtectedResourceSessionTypes[0].resize(3);
+    device->m_ProtectedResourceSessionTypes[1].resize(14);
+    device->m_ProtectedResourceSessionTypes[2].resize(21);
+
+    INIT_FEATURES();
+
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(0), 0);
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(1), 0);
+    EXPECT_EQ(features.ProtectedResourceSessionTypeCount(2), 0);
+}
+
+// 34: Protected Resource Session Types
+// Note: Protected Resource Seesion Type Count must be correctly set for this feature
+// Basic Test
+TEST_F(FeatureSupportTest, ProtectedResourceSessionTypesBasic)
+{
+    device->m_ProtectedResourceSessionTypeCount[0] = 2;
+    device->m_ProtectedResourceSessionTypes[0] = {{1, 1, 2, {3, 5, 8, 13}}, {1, 4, 9, {16, 25, 36, 49}}}; // Some random GUID test data
+
+    INIT_FEATURES();
+
+    EXPECT_EQ(features.ProtectedResourceSessionTypes(), device->m_ProtectedResourceSessionTypes[0]);
+}
+
+// Multicore Test
+TEST_F(FeatureSupportTest, ProtectedResourceSessionTypesMulticore)
+{
+    device->SetNodeCount(2);
+    device->m_ProtectedResourceSessionTypeCount = {2, 1};
+    device->m_ProtectedResourceSessionTypes[0] = {{1, 1, 2, {3, 5, 8, 13}}, {1, 4, 9, {16, 25, 36, 49}}}; // Some random GUID test data
+    device->m_ProtectedResourceSessionTypes[1] = {{5, 7, 9, {11, 13, 15, 17}}};
+
+    INIT_FEATURES();
+
+    EXPECT_EQ(features.ProtectedResourceSessionTypes(0), device->m_ProtectedResourceSessionTypes[0]);
+    EXPECT_EQ(features.ProtectedResourceSessionTypes(1), device->m_ProtectedResourceSessionTypes[1]);
+}
+
+// Unavailable Test
+TEST_F(FeatureSupportTest, ProtectedResourceSessionTypesUnavailable)
+{
+    device->m_ProtectedResourceSessionTypesAvailable = false;
+    device->m_ProtectedResourceSessionTypeCount[0] = 2;
+    device->m_ProtectedResourceSessionTypes[0] = {{1, 1, 2, {3, 5, 8, 13}}, {1, 4, 9, {16, 25, 36, 49}}}; // Some random GUID test data
+
+    INIT_FEATURES();
+
+    // TODO: Decide on expected behavior
+    EXPECT_EQ(features.ProtectedResourceSessionTypes(0).size(), 2);
+    EXPECT_EQ(features.ProtectedResourceSessionTypes()[0], GUID());
+    EXPECT_EQ(features.ProtectedResourceSessionTypes()[1], GUID());
```

Comment:
I wonder what is the expected behavior for ProtectedResourceSessionTypes if the query fails. Right now it will still initialize the returned array (vector) with the size returned from ProtectedResourceSessionTypeCount and each entry would be an empty GUID object.

Reply:
Good question!  Returning an empty vector might be better in this case?

Reply:
I agree. Returning an empty vector is now the default behavior of this API.